File: crates/api/src/http/dto/response/pool_history.rs
//! Wire shape for one hourly bucket of a pool's activity history.
//!
//! Mirrors `yog_core::domain::PoolHistoryBucket`, with one
//! presentation-derived field (`effectiveFeeBps`) computed by the **same
//! function** as on `PoolResponse` — `pool::effective_fee_bps`, which this
//! module used to duplicate inline. Returned as a plain ordered array (oldest →
//! newest) by `GET /api/pools/{address}/history` — chart-ready, no pagination
//! (the window is bounded by `days`).
//!
//! `lpFeesUsd` used to be derived here too, as `fees - protocol`. It is now
//! read from the domain type: that formula credited the referral to the LPs,
//! and it was written twice — here and on `PoolResponse`. The split is defined
//! once, in `meteora_damm_v2_pool_hourly_activity` (`.project` ticket 05).

use chrono::{DateTime, Utc};
use rust_decimal::Decimal;
use serde::Serialize;

use yog_core::domain::PoolHistoryBucket;

use super::pool::effective_fee_bps;

#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub(crate) struct PoolHistoryBucketResponse {
    pub(crate) bucket: DateTime<Utc>,
    pub(crate) volume_usd: Option<Decimal>,
    pub(crate) fees_usd: Option<Decimal>,
    pub(crate) protocol_fees_usd: Option<Decimal>,
    pub(crate) referral_fees_usd: Option<Decimal>,
    pub(crate) lp_fees_usd: Option<Decimal>,
    pub(crate) effective_fee_bps: Option<Decimal>,
    pub(crate) liquidity_added_usd: Option<Decimal>,
    pub(crate) liquidity_removed_usd: Option<Decimal>,
    pub(crate) fees_claimed_usd: Option<Decimal>,
    pub(crate) rewards_claimed_usd: Option<Decimal>,
    pub(crate) swap_count: Option<i64>,
}

impl From<PoolHistoryBucket> for PoolHistoryBucketResponse {
    fn from(b: PoolHistoryBucket) -> Self {
        Self {
            bucket: b.bucket,
            volume_usd: b.volume_usd,
            fees_usd: b.fees_usd,
            protocol_fees_usd: b.protocol_fees_usd,
            referral_fees_usd: b.referral_fees_usd,
            lp_fees_usd: b.lp_fees_usd,
            effective_fee_bps: effective_fee_bps(b.fees_usd, b.volume_usd),
            liquidity_added_usd: b.liquidity_added_usd,
            liquidity_removed_usd: b.liquidity_removed_usd,
            fees_claimed_usd: b.fees_claimed_usd,
            rewards_claimed_usd: b.rewards_claimed_usd,
            swap_count: b.swap_count,
        }
    }
}

#[cfg(test)]
#[path = "tests/pool_history_tests.rs"]
mod tests;

File: web/src/lib/api/schema/pool-history.ts
import * as z from "zod";
import { BigDecimal, Rfc3339 } from "./shared";

// ─────────────────────────────────────────────────────────────────────
// PoolHistory — mirrors `api::http::dto::response::PoolHistoryBucketResponse`
// ─────────────────────────────────────────────────────────────────────

/**
 * One hourly bucket of a pool's activity history, as returned by
 * `GET /api/pools/{address}/history?days=N`.
 *
 * Rust side (api/src/http/dto/response/pool_history.rs), camelCase:
 *
 * ```rust
 * #[serde(rename_all = "camelCase")]
 * pub struct PoolHistoryBucketResponse {
 *     bucket: DateTime<Utc>,
 *     volume_usd: Option<Decimal>,
 *     fees_usd: Option<Decimal>,
 *     protocol_fees_usd: Option<Decimal>,
 *     referral_fees_usd: Option<Decimal>,
 *     lp_fees_usd: Option<Decimal>,
 *     effective_fee_bps: Option<Decimal>,
 *     liquidity_added_usd: Option<Decimal>,
 *     liquidity_removed_usd: Option<Decimal>,
 *     fees_claimed_usd: Option<Decimal>,
 *     rewards_claimed_usd: Option<Decimal>,
 *     swap_count: Option<i64>,
 * }
 * ```
 *
 * USD metrics arrive as precision-safe decimal strings (`BigDecimal`),
 * null when no priced activity of that kind happened in the bucket.
 * `protocolFeesUsd`, `referralFeesUsd` and `lpFeesUsd` are the three shares
 * of `feesUsd`, summing back to it exactly — see `pool.ts` for why
 * `lpFeesUsd` must be read rather than derived from the other two.
 * `swapCount` is a plain JSON number. The endpoint returns the buckets as
 * an ordered array (oldest → newest), not a paginated page.
 */
export const PoolHistoryBucketSchema = z.object({
  bucket: Rfc3339,
  volumeUsd: BigDecimal.nullable(),
  feesUsd: BigDecimal.nullable(),
  protocolFeesUsd: BigDecimal.nullable(),
  referralFeesUsd: BigDecimal.nullable(),
  lpFeesUsd: BigDecimal.nullable(),
  effectiveFeeBps: BigDecimal.nullable(),
  liquidityAddedUsd: BigDecimal.nullable(),
  liquidityRemovedUsd: BigDecimal.nullable(),
  feesClaimedUsd: BigDecimal.nullable(),
  rewardsClaimedUsd: BigDecimal.nullable(),
  swapCount: z.number().int().nullable(),
});

export const PoolHistorySchema = z.array(PoolHistoryBucketSchema);

export type PoolHistoryBucketResponse = z.infer<typeof PoolHistoryBucketSchema>;
export type PoolHistoryResponse = z.infer<typeof PoolHistorySchema>;
